library/alloc/src/boxed/thin.rs

```diff
@@ -91,6 +93,87 @@ impl<T> ThinBox<T> {
 
 #[unstable(feature = "thin_box", issue = "92791")]
 impl<Dyn: ?Sized> ThinBox<Dyn> {
+    #[cfg(not(no_global_oom_handling))]
+    fn new_unsize_zst<T>(value: T) -> Self
+    where
+        T: Unsize<Dyn>,
+    {
+        assert!(mem::size_of::<T>() == 0);
+
+        #[repr(C)]
+        struct ReprC<A, B> {
+            a: A,
+            b: B,
+        }
+
+        struct EmptyArray<T> {
+            _array: [T; 0],
+        }
+
+        unsafe impl<T> Freeze for EmptyArray<T> {}
```

Comment:
```suggestion
        // SAFETY: this is a zero-sized type, there can't be any mutable memory here.
        // It's a private type so this can never leak to the user either.
        // If the user tried to write through the shared reference to this type,
        // they would be causing library UB as that's a write outside the memory
        // inhabited by their type (which is zero-sized). Therefore making this
        // language UB is justified.
        unsafe impl<T> Freeze for EmptyArray<T> {}
```